utils.go

```diff
@@ -1,87 +1,87 @@
 	"github.com/disintegration/imaging" // go get -u github.com/disintegration/imaging
 	"github.com/rwcarlsen/goexif/exif"  // go get -u github.com/rwcarlsen/goexif/exif
+	"github.com/sirupsen/logrus"        // go get -u github.com/sirupsen/logrus
 
-	"errors"
-	"image/gif"
-	"image/jpeg"
-	"image/png"
-	"strings"
+	"bytes"
 )
 
 // ReadImage makes a copy of image (jpg,png or gif) and applies
 // all necessary operation to reverse its orientation to 1
 // The result is a image with corrected orientation and without
 // exif data.
-func ReadImage(fpath string) *image.Image {
-	var img image.Image
-	var err error
-	// deal with image
-	ifile, err := os.Open(fpath)
-	if err != nil {
-		logrus.Warnf("could not open file for image transformation: %s", fpath)
-		return nil
-	}
-	defer ifile.Close()
-	filetype, err := GetSuffix(fpath)
-	if err != nil {
-		return nil
-	}
-	if filetype == "jpg" {
-		img, err = jpeg.Decode(ifile)
-		if err != nil {
-			return nil
-		}
-	} else if filetype == "png" {
-		img, err = png.Decode(ifile)
-		if err != nil {
-			return nil
-		}
-	} else if filetype == "gif" {
-		img, err = gif.Decode(ifile)
-		if err != nil {
-			return nil
-		}
-	}
+func ReadImage(imgBody []byte) *[]byte {
+	imgBodyReader := bytes.NewReader(imgBody)
 	// deal with exif
-	efile, err := os.Open(fpath)
-	if err != nil {
-		logrus.Warnf("could not open file for exif decoder: %s", fpath)
+	var err error
+	img, imgExtension, err := image.Decode(imgBodyReader)
+	if imgExtension != "png" && imgExtension != "jpg" && imgExtension != "jpeg" && imgExtension != "gif" {
+		fmt.Printf("image type %s has no exif to check for orientation", imgExtension)
+		return &imgBody
 	}
-	defer efile.Close()
-	x, err := exif.Decode(efile)
+	x, err := exif.Decode(imgBodyReader)
 	if err != nil {
 		if x == nil {
-			// ignore - image exif data has been already stripped
+			fmt.Printf("image has no exif data, no further exif manipulation is needed")
+			return &imgBody
 		}
-		logrus.Errorf("failed reading exif data in [%s]: %s", fpath, err.Error())
+		logrus.Errorf("failed reading exif data: %s", err.Error())
 	}
 	if x != nil {
 		orient, _ := x.Get(exif.Orientation)
 		if orient != nil {
-			logrus.Infof("%s had orientation %s", fpath, orient.String())
+			if orient == 1 {
+				fmt.Printf("image already has correct orientation, no further exif manipulation is needed")
+				return &imgBody
+			}
+			logrus.Infof("image had orientation %s", orient.String())
 			img = reverseOrientation(img, orient.String())
+			switch imgExtension {
+			case "png":
+				buffer := new(bytes.Buffer)
+				err := png.Encode(buffer, img)
+				if err != nil {
+					fmt.Printf("error while encoding corrected image: %s", err)
+				}
+				imgBody = buffer.Bytes()
+				return &imgBody
+			case "gif":
+				buffer := new(bytes.Buffer)
+				err := gif.Encode(buffer, img, nil)
+				if err != nil {
+					fmt.Printf("error while encoding corrected image: %s", err)
```

Comment:
I know it was originally like that, but it's not good practice for a library to print to standard output.
We could receive the logger as a parameter. :)

Reply:
Like this:
```go
func ReadImage(imgBody []byte, logger *logrus.Entry) {
```

Reply:
Kage-bunshin's logger also uses `logrus` under the hood, so we would be able to do something like this:
```go
rotatedImage := ReadImage(imageBytes, kage.Logger.Entry)
```

Reply:
Okay that makes sense, ill redo these logs asap